src/ray/raylet/scheduling/cluster_resource_scheduler.h

```diff
@@ -83,16 +85,22 @@ class ClusterResourceScheduler {
   ///  prioritize_local_node if set to true.
   ///  \param requires_object_store_memory: take object store memory usage as part of
   ///  scheduling decision.
-  ///  \param is_infeasible[out]: It is set true if the task is not schedulable because it
-  ///  is infeasible.
+  ///  \param is_infeasible[out]: It is set
+  ///  true if the task is not schedulable because it is infeasible.
+  ///  \param forward_to: The node to be forwarded to for scheduling.
+  ///  \param use_required_resources: Whether using required_resources (instead of
+  ///  required_placement_resources) for scheduling.
   ///
   ///  \return emptry string, if no node can schedule the current request; otherwise,
   ///          return the string name of a node that can schedule the resource request.
-  scheduling::NodeID GetBestSchedulableNode(const TaskSpecification &task_spec,
-                                            bool prioritize_local_node,
-                                            bool exclude_local_node,
-                                            bool requires_object_store_memory,
-                                            bool *is_infeasible);
+  scheduling::NodeID GetBestSchedulableNode(
+      const TaskSpecification &task_spec,
+      bool prioritize_local_node,
+      bool exclude_local_node,
+      bool requires_object_store_memory,
+      bool *is_infeasible,
+      scheduling::NodeID forward_to = scheduling::NodeID::Nil(),
+      bool use_required_resources = false);
```

Comment:
This is also a leaky abstraction. Could TaskSpecification has a method to tell if it requires placement group resource or normal resource?

Reply:
ideally both `forward_to` and `use_required_resources` should not be part of public apis.

Reply:
Having `use_required_resources` is tricky. For an actor like this:
`@remote()
 class TestedActor:`
Its required_placement_resources would be ("CPU": 1) while required_resources would be empty. When the raylet scheduler schedules such an actor, what it previously did is choosing a random node inside `gcs_actor_scheduler`. So now if we want `cluster_task_manager` for queueing, then we need this parameter to tell us (when dequeueing) to use the **empty** required_resources such that the random_scheduling_policy will be enforced. That will keep the same scheduling logic with the one before.

Reply:
This `use_required_resources` can be removed once we unify the required_placement_resources and required_resources. @scv119

Reply:
Yes, I'll unify it as part of 2.0 so it's just temporary.